### ISSUE TITLE
Require newline between Javadoc contents and at-clauses

### ISSUE THREAD

**Issue by mkordas** [yegor256/qulice#708]
Both [Google Style Guide](https://google.github.io/styleguide/javaguide.html#s7.1.2-javadoc-paragraphs) and [Stephen Colebourne](http://blog.joda.org/2012/11/javadoc-coding-standards.html) recommend to put newline before the group of "at-clauses" if they are present.

In our code we have various approaches, I see some people adding them why others are removing.

According to https://github.com/teamed/qulice/issues/708#issuecomment-184518026, we should implement rule to require newline based on a Javadoc length.

This should be invalid:

``` java
  /**
   * Javadoc text.
   * 
   * @return the baz content, null if not processed
   */
  public String process() {...}
```

``` java
  /**
   * Javadoc text.
   *
   * Javadoc text.
   * @return the baz content, null if not processed
   */
  public String process() {...}
```

While this is correct:

``` java
  /**
   * Javadoc text.
   * @return the baz content, null if not processed
   */
  public String process() {...}
```

``` java
  /**
   * Javadoc text.
   *
   * Javadoc text.
   *
   * @return the baz content, null if not processed
   */
  public String process() {...}
```

**Comment by krzyk** [comment]
@mkordas I'm in the "remove that empty line" crowd :) Unless it is a class comment. So let's see what @yegor256 thinks. Should we enforce an empty line between method description and at-clauses always?

**Comment by mkordas** [comment]
@krzyk yes, let's make one decision and be consistent. Either enforcing or prohibiting this newline works for me

**Comment by yegor256** [comment]
@mkordas @krzyk I use this rule in all code I write: if there is only one paragraph/line before tags, I don't put empty line. In all other situations I add an empty line. Examples:

```
/**
 * Just one line here.
 * @since 0.1
 */
/**
 * First line.
 *
 * Second par.
 *
 * @since 0.1
 */
```

I believe it's logical.

**Comment by mkordas** [comment]
@yegor256 makes sense, thanks
@krzyk do you agree? should I update description accordingly?

**Comment by krzyk** [comment]
@mkordas yes

**Comment by mkordas** [comment]
@krzyk updated

**Comment by krzyk** [comment]
@davvd valid bug

**Comment by davvd** [comment]
> @davvd valid bug

@krzyk I just added bug tag here

**Comment by krzyk** [comment]
@davvd this is postponed

**Comment by davvd** [comment]
> @davvd this is postponed

@krzyk right, I added "postponed" label

**Comment by davvd** [comment]
> @davvd this is postponed

@krzyk someone else will help in this task, no problem at all

**Comment by davvd** [comment]
@mkordas thanks for this report, I added 15 mins to your account, in transaction `56e291bc9e702429ce00032e`

**Comment by yegor256** [comment]
Implemented as discussed in this thread in commit 7ea4b620c.

Added a new Checkstyle rule `JavadocEmptyLineBeforeTagCheck` that follows the convention proposed here:

- If the Javadoc body before at-clauses is a **single paragraph**, an empty Javadoc line between the body and the first at-clause is **not allowed**.
- If the body is **multi-paragraph** (separated by empty Javadoc lines), an empty Javadoc line before the first at-clause is **required**.

The check is wired into the default `checks.xml`, covered by `ChecksTest/JavadocEmptyLineBeforeTagCheck/{Valid,Invalid}.java` with a matching `violations.txt`, and Javadocs across the repo have been normalized to satisfy it. All 285 tests pass.
